flink-yarn/src/main/java/org/apache/flink/yarn/YarnApplicationFileUploader.java

```diff
@@ -163,7 +164,15 @@ YarnLocalResourceDescriptor registerSingleLocalResource(
         addToRemotePaths(whetherToAddToRemotePaths, resourcePath);
 
         if (Utils.isRemotePath(resourcePath.toString())) {
-            final FileStatus fileStatus = fileSystem.getFileStatus(resourcePath);
+            final FileSystem srcFs = resourcePath.getFileSystem(fileSystem.getConf());
+            final FileStatus fileStatus;
+            if (srcFs.getScheme().equals(fileSystem.getScheme())) {
+                fileStatus = srcFs.getFileStatus(resourcePath);
+            } else {
+                Tuple2<Path, Long> remoteFileInfo =
+                        uploadLocalFileToRemote(resourcePath, relativeDstPath, false);
```

Comment:
1. Why do we need to copy the file from another remote filesystem to the default remote filesystem?
2. The method name `uploadLocalFileToRemote` is misleading. The file being uploaded is not local.
3. `uploadLocalFileToRemote` internally uses `o.a.hadoop.fs.FileUtil.copy()`. Looking into hadoop source codes, this util method opens the src/dst files as `Input/OutputStream` and copy the bytes. That means all the data go through the local client.